consensus/hotstuff/committees/consensus_committee.go

```diff
@@ -328,112 +283,100 @@ func (c *Consensus) TimeoutThresholdForView(view uint64) (uint64, error) {
 	ready()
 
 	for {
 		select {
 		case <-ctx.Done():
 			return
-		case block := <-c.committedEpochsCh:
-			epoch := c.state.AtBlockID(block.ID()).Epochs().Next()
-			_, err := c.prepareEpoch(epoch)
-			if err != nil {
-				ctx.Throw(err)
-			}
-		case <-c.epochFallbackTriggeredCh:
-			err := c.onEpochFallbackModeTriggered()
+		case handleEvent := <-c.events:
+			err := handleEvent()
 			if err != nil {
 				ctx.Throw(err)
 			}
 		}
 	}
 }
 
-// EpochCommittedPhaseStarted informs the `committee.Consensus` that the block starting the Epoch Committed Phase has been finalized.
+// EpochCommittedPhaseStarted informs `committees.Consensus` that the first block in flow.EpochPhaseCommitted has been finalized.
+// This event consumer function enqueues an event handler function for the single event handler thread to execute.
 func (c *Consensus) EpochCommittedPhaseStarted(_ uint64, first *flow.Header) {
-	c.committedEpochsCh <- first
+	c.events <- func() error {
+		return c.handleEpochCommittedPhaseStarted(first)
+	}
 }
 
-// EpochFallbackModeTriggered passes the protocol event to the worker thread.
-func (c *Consensus) EpochFallbackModeTriggered(uint64, *flow.Header) {
-	c.epochFallbackTriggeredCh <- struct{}{}
+// EpochExtended informs `committees.Consensus` that a block including a new epoch extension has been finalized.
+// This event consumer function enqueues an event handler function for the single event handler thread to execute.
+func (c *Consensus) EpochExtended(_ uint64, refBlock *flow.Header, _ flow.EpochExtension) {
+	c.events <- func() error {
+		return c.handleEpochExtended(refBlock)
+	}
 }
 
-// onEpochFallbackModeTriggered handles the protocol event for epoch fallback mode being triggered.
-// When this occurs, we inject a fallback epoch to the committee which extends the current epoch.
-// This method must also be called on initialization, if epoch fallback mode was triggered in the past.
+// handleEpochExtended executes all state changes required upon observing an EpochExtended event.
+// This function conforms to eventHandlerFunc.
+// When an extension is observed, we re-compute leader selection for the current epoch, taking into
+// account the most recent extension (included as of refBlock).
 // No errors are expected during normal operation.
-func (c *Consensus) onEpochFallbackModeTriggered() error {
-	// we respond to epoch fallback being triggered at most once, therefore
-	// the core logic is protected by an atomic bool.
-	// although it is only valid for epoch fallback to be triggered once per spork,
-	// we must account for repeated delivery of protocol events.
-	if !c.isEpochFallbackHandled.CompareAndSwap(false, true) {
-		return nil
-	}
-
-	currentEpochCounter, err := c.state.Final().Epochs().Current().Counter()
+func (c *Consensus) handleEpochExtended(refBlock *flow.Header) error {
+	currentEpoch := c.state.AtHeight(refBlock.Height).Epochs().Current()
```

Comment:
This is from @durkmurder's comment here: https://github.com/onflow/flow-go/pull/6154#discussion_r1662939153, for the reason you guessed. Using `Height` also saves an ID computation. 

You're right that `state.AtHeight` always hits the database. We also have the method `BlockIDByHeight` which uses the `height->blockID` cache in `storage.Headers`. If that is straightforward to plug in, I'll stick with `Height`, otherwise will revert back to `BlockID`.

Reply:
I replaced the explicit database call with a call to `storage.Headers`, which will check the cache first, in https://github.com/onflow/flow-go/pull/6154/commits/62c5d690bfc614c86a165a7ab8cf507a7a55fd12.